request_without_range.go

```diff
@@ -0,0 +1,56 @@
+package pget
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+
+	"github.com/cheggaaa/pb/v3"
+)
+
+func DownloadFiles(urls []string, dest string) error {
+	for _, link := range urls {
+		if err := DownloadFile(link, dest, context.Background()); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+func DownloadFile(url string, dest string, ctx context.Context) error {
```

Comment:
This one also
https://pkg.go.dev/context#:~:text=The%20Context%20should%20be%20the%20first%20parameter%2C%20typically%20named%20ctx

And Could you please move this function into the `Download` function which already exists.

Reply:
@hellojukay Please do this one 🙏 

> Could you please move the content of this function into the `Download` function which already exists? If you can move, I think `DownloadFiles` function will be able to remove.

Basically, we want it to be completed in the [Run](https://github.com/Code-Hex/pget/blob/40ee37bb3515f9cae7c8155a95068b49042aeee1/pget.go#L41) method.

Describe what you want to be done.

1. Please modify [these processes](https://github.com/Code-Hex/pget/blob/40ee37bb3515f9cae7c8155a95068b49042aeee1/pget.go#L58-L73) so that they are done before the call to function `Check`
2. There is already a Download function, so I want you to use that. And please add an [option](https://github.com/Code-Hex/pget/blob/40ee37bb3515f9cae7c8155a95068b49042aeee1/pget.go#L75-L78) to enable range-request downloads. I think the definition would be something like this.

```go
func AsRangeRequest() DownloadOption {
	return func(c *DownloadConfig) {
		c.makeRequestOption.useRangeRequest = true
	}
}

// Use if normal download
opts := []DownloadOption{
	WithUserAgent(pget.useragent),
	WithReferer(pget.referer),
}

// Use When Parallel-Download
opts := []DownloadOption{
	WithUserAgent(pget.useragent),
	WithReferer(pget.referer),
	AsRangeRequest(),
}
```

3. fix error handle for `ErrNotSupportRequestRange` at this line.
https://github.com/Code-Hex/pget/blob/40ee37bb3515f9cae7c8155a95068b49042aeee1/pget.go#L54-L56

The code will be like below

```go
if err != nil {
	if errors.Is(err, ErrNotSupportRequestRange) {
		return Download(ctx, &DownloadConfig{
			Filename:      filename,
			Dirname:       dir,
			ContentLength: target.ContentLength,
			Procs:         pget.Procs,
			URLs:          target.URLs,
			Client:        client,
		},  WithUserAgent(pget.useragent), WithReferer(pget.referer))
	}
	return err
}
```

Reply:
if where happen error, target is nil , so we can not know ContentLength 
```
			ContentLength: target.ContentLength,
			URLs:          target.URLs,

```
So we have to modify the Download function to match the contentLength as 0 or modify the check function to get the contentLenght

Reply:
Ah, yes. Please use pget.URLs instead of it

Reply:
how about this 
```go
var dir string
if pget.Output != "" {
	fi, err := os.Stat(pget.Output)
	if err == nil && fi.IsDir() {
		dir = pget.Output
	} else {
		dir, _ = filepath.Split(pget.Output)
		if dir != "" {
			if err := os.MkdirAll(dir, 0755); err != nil {
				return errors.Wrapf(err, "failed to create diretory at %s", dir)
			}
		}
	}
}

target, err := Check(ctx, &CheckConfig{
	URLs:    pget.URLs,
	Timeout: time.Duration(pget.timeout) * time.Second,
	Client:  client,
})
if err != nil {
	if errors.Is(err, ErrNotSupportRequestRange) {
		return pget.downloadFilesWithoutRange(context.Background(), pget.URLs, dir)
	}
	return err
}

```

Reply:
I'm sorry its not acceptable. I indicated a way to use optional function `AsRangeRequest`.
Please follow this way 🙏